src/udp.rs

```diff
@@ -39,103 +114,293 @@ impl UdpSocket {
-                {
-                    self.socket
-                        .reset_interrupt(bus, socketn::Interrupt::Timeout)?;
-                    return Err(NbError::Other(UdpSocketError::WriteTimeout));
-                }
+            match self.try_flush_tx(bus) {
+                Err(NbError::WouldBlock) => {}
+                result => return result,
             }
         }
     }
 
+    fn try_flush_tx<SpiBus: Bus>(
+        &self,
+        bus: &mut SpiBus,
+    ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
+        let tx_read = self.socket.get_tx_read_pointer(bus)?;
+        let tx_write = self.socket.get_tx_write_pointer(bus)?;
+
+        if tx_read == tx_write {
+            if self.socket.has_interrupt(bus, socketn::Interrupt::SendOk)? {
+                self.socket
+                    .reset_interrupt(bus, socketn::Interrupt::SendOk)?;
+
+                return Ok(());
+            } else if self
+                .socket
+                .has_interrupt(bus, socketn::Interrupt::Timeout)?
+            {
+                self.socket
+                    .reset_interrupt(bus, socketn::Interrupt::Timeout)?;
+
+                return Err(NbError::Other(UdpSocketError::WriteTimeout));
+            }
+            // other interrupts are unreachable in UDP mode.
+
+            return Ok(());
+        }
+
+        Err(NbError::WouldBlock)
+    }
+
     /// Sets a new destination before performing the send operation.
+    ///
+    /// # Returns
+    ///
+    /// The amount of bytes that were sent. The caller should make sure to
+    /// check and send the rest of the `send_buffer` data.
     fn socket_send_to<SpiBus: Bus>(
         &mut self,
         bus: &mut SpiBus,
         remote: SocketAddrV4,
         send_buffer: &[u8],
     ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
         self.set_destination(bus, remote)?;
-        self.socket_send(bus, send_buffer)
+
+        self.socket_send_all(bus, send_buffer)
     }
 
     /// Receive data and mutate the `receive_buffer`.
     ///
+    /// `receive_buffer` will only be used for receiving the packet data.
+    /// The [`UdpHeader`]'s - [`SocketAddrV4`] and `len` will be read separately from the buffer.
+    /// Note that the header is part of the internal RX buffer so `receive_buffer` can be smaller
+    /// in size by 8 bytes.
+    ///
+    /// If the packet len is larger than the internal RX buffer, the data will be truncated.
+    ///
     /// If [`Interrupt::Receive`] is not set, it will always return [`NbError::WouldBlock`].
+    ///
+    ///
+    /// Packet frame, as described in W5200 docs section 5.2.2.1:
+    ///
+    /// ```text
+    /// |<-- read_pointer                                 read_pointer + received_size -->|
+    /// | Destination IP Address | Destination Port | Byte Size of DATA | Actual DATA ... |
+    /// |    --- 4 Bytes ---     |  --- 2 Bytes --- |  --- 2 Bytes ---  |      ....       |
+    /// ```
```

Comment:
```suggestion
```

This data isn't relevant here, and is documented elsewhere

Reply:
Removed